### ISSUE TITLE
Paper support(i don't know if has)?

### ISSUE THREAD

**Issue by yurinogueira** [MockBukkit/MockBukkit#143]
![image](https://user-images.githubusercontent.com/55946612/101438720-c519ac80-38f1-11eb-9407-68b34ef4f689.png)

**Comment by JacksonChen666** [comment]
I can see no relation to this repo, because it is created for testing plugins and not included in the jar file (if not specified).

**Comment by seeseemelk** [comment]
Paper is currently not supported, although I see no reason support for it couldn't be added. But until such support is added, you should try to only use Bukkit APIs.

If you absolutely need to use Paper functions, you could use the same solution used for running NMS code with MockBukkit #139. This solution involves wrapping the Paper functions in a interface, and passing an implementing class in via the constructor. During testing, an alternative implementation of that interface can be provided which stubs out those functions.

**Comment by Narimm** [comment]
I have a branch for this ...but its proving a nightmare to use Paper due to all the Static loads that paper usings in the SimplePluginManager.

**Comment by Narimm** [comment]
Also Paper is now providing a stack of methods from Adventure - I would suggest this be closed as label:wontfix

**Comment by MiniDigger** [comment]
If mockbukkit would compile against paper and stub/implement the missing methods, it would still be usable for spigot plugins, right? As spigot plugins work against the paper api without issues.
It ppl agree, I would be willing to change the compile target and get it compile clean.

**Comment by TheBusyBiscuit** [comment]
@MiniDigger It should still be usable for spigot plugins, yes.
However if we were to target the `paper-api`, We could also just have a dedicated `MockBukkit/MockPaper` repository for it just incase.

Right now one of our biggest obstacles are ever-changing method signatures in the Spigot-API, thanks to it's `-SNAPSHOT` versioning scheme. Every couple of days, method signatures change and we need to add new stubs or edit existing ones.
Targeting an exact version also seems impossible, as spigot's repository auto-purges older jars under the same snapshot.

Not sure if Paper's repository supports specific versions to be built against.
But given that Paper's development progresses much faster than Spigot's, this problem will likely scale up.
Eventually we will probably need to create an automatic "stub and fix" workflow which we can put on a cron job. Or perhaps you have a better idea how to tackle this issue?

Anyway, for now targeting the paper-api is probably not in reach. 
Not until this problem has been addressed at least.

> I would suggest this be closed as label:wontfix

@Narimm
I do wanna keep the issue open at the moment since a dedicated Paper mock (or switching completely to Paper-API for when Spigot continues to fade) is something I am generally open to.
It is a legtimate request and Paper has also taken the lead already. (👀 https://bstats.org/global/bukkit#serverSoftware)

**Comment by MiniDigger** [comment]
> Not sure if Paper's repository supports specific versions to be built against.

pretty sure we recently disabled auto purge, our repo should keep snapshots indefinitely (or for two years, I can't remember what we decided) now. 

I don't know exactly how MockBukkit works, but maybe it would be possible to have a MockPaper project that is able to extend the MockBukkit classes and just adds the methods paper adds? 

Traditional mock apis work on bytecode level, to circumvent the fact that you have to implement every method, even tho you only want to mock a simple one. maybe it would be more future proof to implement mockbukkit as a list of when-then calls in mockito or something? 

> It is a legtimate request and Paper has also taken the lead already

to further prove that point home: historical bstats data for this: https://kennytv.eu/secretgraph/

**Comment by TheBusyBiscuit** [comment]
> pretty sure we recently disabled auto purge, our repo should keep snapshots indefinitely (or for two years, I can't remember what we decided) now.

Hmm, that would certainly help. Could you perhaps confirm this? If so, then I will definitely look into either changing the target to `paper-api` or creating a `MockPaper` repository for it as a prototype. If that would help solve our current stub problem, then that would be a hard argument in favour of it, two years is plenty of time to manually bump the version inbetween.
Though I will have to discuss this with @seeseemelk internally.

> I don't know exactly how MockBukkit works, but maybe it would be possible to have a MockPaper project that is able to extend the MockBukkit classes and just adds the methods paper adds?

It could be possible, though I am not 100% sure, will have to see what the best approach would be.
Thanks for your input though!

**Comment by seeseemelk** [comment]
> to further prove that point home: historical bstats data for this: https://kennytv.eu/secretgraph/

This graph is really interesting to me. Given that it seems that Paper is being used more that Spigot, I'd definitely say that we should perhaps prioritize Paper over pure Spigot.


>> pretty sure we recently disabled auto purge, our repo should keep snapshots indefinitely (or for two years, I can't remember what we decided) now.
>
> Hmm, that would certainly help. Could you perhaps confirm this?

I would also like to know this.

**Comment by ItsLiyua** [comment]
Isn't Bukkit outdated anyway? Like? Why use Bukkit when you can use Spigot? Why use Spigot when you can use Paper?

**Comment by rona-tombo** [comment]
is there any update?

I'm working on a library that relies on the paper API, so I'm really looking forward to it!
If necessary, I can create a PR to replace the dependency with paper-api.

**Comment by rona-tombo** [comment]
https://papermc.io/repo/#browse/browse:maven-snapshots:io%2Fpapermc%2Fpaper%2Fpaper-api%2F1.18.1-R0.1-SNAPSHOT%2F1.18.1-R0.1-20211219.173229-16

It looks like the snapshot of paper is still intact, so you can build for a specific version by specifying the following
```groovy
dependencies {
    api "io.papermc.paper:paper-api:1.18.1-R0.1-20211219.173229-16"
}
```
(And perhaps we should use `implementation` instead of `api`)

**Comment by Insprill** [comment]
Has any more thought been put into this? I would love to see it and right now there are actually some issues with the Adventure API, specifically when calling the PlayerJoinEvent when adding a player. As Paper grows and more developers are building plugins around it, this would be great to have as long as specific builds can be targeted.

**Comment by thelooter** [comment]
I think this Issue can now be closed since Paper Support has been added

**Comment by JacksonChen666** [comment]
> I think this Issue can now be closed since Paper Support has been added

related issue appears to be #347

**Comment by TheBusyBiscuit** [comment]
`MockBukkit` officially uses the Paper API now, as of #347 (release v2.0.0 +)
